### PR TITLE
ARTEMIS-2251 Large messages might not be deleted when server crashed

### PR DESCRIPTION
**Pull request by qihongxu**
When deleting large messages, artemis will use storePendingLargeMessage to insert a temporary record in journal for reload, in case server crashed and large messages stayed forever. But in storePendingLargeMessage that appendAddRecord inserts records asynchronously. In this way there are potential risks that tasks in executor get lost due to server crash, which may lead to undeletable large messages. To solve this problem a Boolean is added to storePendingLargeMessage so that it will be forced to use SimpleWaitIOCallback in delete situation.

### CODE REVIEW

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/impl/journal/JournalStorageManager.java [hunk 1]
```diff
@@ -432,7 +432,7 @@ public void freeDirectBuffer(ByteBuffer buffer) {
       journalFF.releaseBuffer(buffer);
    }
 
-   public long storePendingLargeMessage(final long messageID, long recordID) throws Exception {
```
Comment (clebertsuconic) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/impl/journal/JournalStorageManager.java]: I'm not sure this is a good idea, this is introducing a blocking operation.

Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/impl/journal/JournalStorageManager.java]: which will block any netty processing.

Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/impl/journal/JournalStorageManager.java]: although delete itself is already blocking.. so I think it's fine.

Comment (michaelandrepearce) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/impl/journal/JournalStorageManager.java]: There is diff between a memory block and an waiting on disk io.

Can this not be async with a callback with a response later?

Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/impl/journal/JournalStorageManager.java]: I am sending an alternate PR.

Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/impl/journal/JournalStorageManager.java]: @qihongxu I fixed it without a blocking operation through a different PR. Please let me know if you see any issues.

Thanks

@michaelandrepearce  ^^

**Pull request closed without merging**